# Questão 1

Uma matriz $H \in R^{n×n}$ é chamada matriz de Hilbert quando

$h_{ij} = \frac{1}{i + j − 1}, \forall i, j = 1, . . . , n.$

Usando o comando `hilbert` da biblioteca `scipy.linalg` em Python, podemos gerar facilmente a matriz de Hilbert de dimensão $n \times n$.

Por exemplo, os comandos

```
from scipy.linalg import hilbert
H = hilbert(3)
```


fornecem a matriz de Hilbert de dimensão $3 \times 3$.
Gere a matriz de Hilbert considerando $n = 3$, $n = 10$ e $n = 30$ e defina $b = Hu$, em que $u =
[1.0, 1.0, . . . , 1.0]^T$ é um vetor coluna com todas as componentes sendo $1.0$. Resolva o sistema $Hx = b$
usando o método da eliminação de Gauss e compare a solução obtida com $u = [1.0, 1.0, . . . , 1.0]^T$.

Comente sobre os valores obtidos.

In [78]:
import numpy as np
from scipy.linalg import hilbert

In [79]:
def SubstituicaoRegressiva(U, c, n):
    x = np.copy(c)
    for i in range(n-1, -1, -1):
        for j in range(i + 1, n):
            x[i] -= (U[i][j] * x[j])
        x[i] /= U[i][i]
    return x

In [80]:
def EliminacaoGauss(A_in, b_in, n):
    A = np.copy(A_in)
    b = np.copy(b_in)
    for j in range(n):
        for i in range(j + 1, n):
            m = A[i][j]/A[j][j]
            b[i] -= (m * b[j])
            for k in range(j + 1, n):
                A[i][k] -= m * A[j][k]
    return SubstituicaoRegressiva(A, b, n)

Após criar funções que aplicam os algoritmos de Substituição Regressiva e Eliminação de Gauss estudados em aula, automatizei os cálculos e a impressão dos vetores e matrizes por meio da rotina `ResolveQuestao`.

Ela recebe o valor de $n$, invoca sua matriz de Hilbert $H$ e o vetor de valores unitários $u$, calcula $b = H u$ e resolve o sistema $H x = b$, imprimindo $H$, $u$, $b$ e $x$.

In [81]:
def ResolveQuestao(n, printVal=True):
    H = hilbert(n)
    u = np.ones(n)
    b = H @ u
    x =  EliminacaoGauss(H, b, n)
    if (printVal):
        print("\n".join(f"H = {H},u = {u},b = {b}".split(",")))
        print(f"\nx = {x}")
    return x

In [82]:
ResolveQuestao(3)

H = [[1.         0.5        0.33333333]
 [0.5        0.33333333 0.25      ]
 [0.33333333 0.25       0.2       ]]
u = [1. 1. 1.]
b = [1.83333333 1.08333333 0.78333333]

x = [1. 1. 1.]


array([1., 1., 1.])

In [83]:
ResolveQuestao(10)

H = [[1.         0.5        0.33333333 0.25       0.2        0.16666667
  0.14285714 0.125      0.11111111 0.1       ]
 [0.5        0.33333333 0.25       0.2        0.16666667 0.14285714
  0.125      0.11111111 0.1        0.09090909]
 [0.33333333 0.25       0.2        0.16666667 0.14285714 0.125
  0.11111111 0.1        0.09090909 0.08333333]
 [0.25       0.2        0.16666667 0.14285714 0.125      0.11111111
  0.1        0.09090909 0.08333333 0.07692308]
 [0.2        0.16666667 0.14285714 0.125      0.11111111 0.1
  0.09090909 0.08333333 0.07692308 0.07142857]
 [0.16666667 0.14285714 0.125      0.11111111 0.1        0.09090909
  0.08333333 0.07692308 0.07142857 0.06666667]
 [0.14285714 0.125      0.11111111 0.1        0.09090909 0.08333333
  0.07692308 0.07142857 0.06666667 0.0625    ]
 [0.125      0.11111111 0.1        0.09090909 0.08333333 0.07692308
  0.07142857 0.06666667 0.0625     0.05882353]
 [0.11111111 0.1        0.09090909 0.08333333 0.07692308 0.07142857
  0.06666667 0.0625 

array([1.        , 0.99999998, 1.00000048, 0.9999957 , 1.00002028,
       0.99994479, 1.00008985, 0.99991378, 1.00004499, 0.99999016])

In [84]:
ResolveQuestao(30)

H = [[1.         0.5        0.33333333 0.25       0.2        0.16666667
  0.14285714 0.125      0.11111111 0.1        0.09090909 0.08333333
  0.07692308 0.07142857 0.06666667 0.0625     0.05882353 0.05555556
  0.05263158 0.05       0.04761905 0.04545455 0.04347826 0.04166667
  0.04       0.03846154 0.03703704 0.03571429 0.03448276 0.03333333]
 [0.5        0.33333333 0.25       0.2        0.16666667 0.14285714
  0.125      0.11111111 0.1        0.09090909 0.08333333 0.07692308
  0.07142857 0.06666667 0.0625     0.05882353 0.05555556 0.05263158
  0.05       0.04761905 0.04545455 0.04347826 0.04166667 0.04
  0.03846154 0.03703704 0.03571429 0.03448276 0.03333333 0.03225806]
 [0.33333333 0.25       0.2        0.16666667 0.14285714 0.125
  0.11111111 0.1        0.09090909 0.08333333 0.07692308 0.07142857
  0.06666667 0.0625     0.05882353 0.05555556 0.05263158 0.05
  0.04761905 0.04545455 0.04347826 0.04166667 0.04       0.03846154
  0.03703704 0.03571429 0.03448276 0.03333333 0.03225806 0.

array([   0.99999999,    1.00000755,    0.99940244,    1.01688855,
          0.76507463,    2.85147185,   -7.80481721,   26.42334899,
        -39.876216  ,   20.4760174 ,   53.86334464, -112.75163499,
         96.57257109,  -20.6029584 ,  -46.30165072,   61.16865385,
        -24.58398834,   65.55474564, -136.59663657,   70.4311136 ,
         -3.81224387,   38.45158387,   30.53850201,  -77.6578579 ,
         14.55490578,  -59.26055472,  119.85546397,  -24.27612804,
        -41.38960999,   19.39120128])

Como $H x = b = H u$, $H x = H u$. Seguindo, desde que $H$ seja não singular, $x = u$.

Tal resultado esperado ocorreu para $n=3$, contudo isso não foi verdade para $n=10$ e $n=30$.

A média do erro relativo entre as coordenadas do vetor $x$ para $n=10$ foi de $3.1E-5$; e, para $n=30$, $40.5$. Tais valores foram calculados nas células abaixo.

Considerando a dependência apresentada entre o erro e o valor de $n$ e o as propriedades dos métodos utilizados, concluí que o desvio deve ocorrer por causa do erro de aproximação em contas que envolvem ponto flutuante. Ainda, como a resolução do sistema depende do algoritmo de Eliminação de Gauss, são realizadas $O(n^3)$ operações — o que explica o maior desvio para $n=30$, uma vez que mais operações levam a maior propagação do erro.

Em suma, a causa de $x = u$ não ser verdade para todo $n$ é o acúmulo de erros de arrendondamento para operações sucessivas. E, como são realizadas $O(n^3)$ operações, o erro cresce com o aumento de $n$.

In [85]:
x = ResolveQuestao(10, False)
u = np.ones(10)
erro_relativo = np.abs(x - u)/u
sum(erro_relativo)/10

3.1117695386262414e-05

In [86]:
x = ResolveQuestao(30, False)
u = np.ones(30)
erro_relativo = np.abs(x - u)/u
sum(erro_relativo)/30

40.476654657524556

# Questão 2

Determine a fatoração LU, sem pivoteamento, da matriz

$A =
\begin{bmatrix}
4 & −8 & 1 \\
6 & 5 & 7 \\
0 & −10 & −3
\end{bmatrix}
.$


Detalhe as operações efetuadas para determinar os fatores $L$ e $U$.
Observação: Essa questão pode resolvida sem o uso de computadores. Nesse caso, pode-se submeter a
versão digital de um arquivo manuscrito.

In [87]:
def FatoracaoLU(A):
    U = np.copy(A)
    n = A.shape[0]
    L = np.eye(n)

    for j in range(0, n):
        for i in range(j + 1, n):
            L[i][j] = U[i][j]/U[j][j]
            for k in range(j + 1, n):
                U[i][k] -= L[i][j] * U[j][k]
            U[i][j] = 0
            print(f"j = {j} & i = {i}: (instante {i+j})\nL = {L}\nU = {U}\n\n")
    return L, U

In [88]:
A = np.array(
    [[4.0, -8.0, 1.0]
    , [6.0, 5.0, 7.0]
    , [0.0, -10.0, -3.0]])
L, U = FatoracaoLU(A)
print(f"\n\nL = {L}\nU = {U}\n\n")

j = 0 & i = 1: (instante 1)
L = [[1.  0.  0. ]
 [1.5 1.  0. ]
 [0.  0.  1. ]]
U = [[  4.   -8.    1. ]
 [  0.   17.    5.5]
 [  0.  -10.   -3. ]]


j = 0 & i = 2: (instante 2)
L = [[1.  0.  0. ]
 [1.5 1.  0. ]
 [0.  0.  1. ]]
U = [[  4.   -8.    1. ]
 [  0.   17.    5.5]
 [  0.  -10.   -3. ]]


j = 1 & i = 2: (instante 3)
L = [[ 1.          0.          0.        ]
 [ 1.5         1.          0.        ]
 [ 0.         -0.58823529  1.        ]]
U = [[ 4.         -8.          1.        ]
 [ 0.         17.          5.5       ]
 [ 0.          0.          0.23529412]]




L = [[ 1.          0.          0.        ]
 [ 1.5         1.          0.        ]
 [ 0.         -0.58823529  1.        ]]
U = [[ 4.         -8.          1.        ]
 [ 0.         17.          5.5       ]
 [ 0.          0.          0.23529412]]




In [89]:
L@U - A

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

A estratégia para fatoração LU é, inicializados $L = I$ e $U = A$, realizar operações elementares de subtração entre uma linha e o múltiplo $m$ de outra linha. Isso deve ocorrer de forma que, coluna a coluna, os elementos abaixos da diagonal principal de $U$ sejam nulos e os elementos $L[i][j]$ abaixo da diagonal principal de $L$ (condição $i > j$) contenham os múltiplos necessários para anular o respectivo elemento $U[i][j]$. 

Expressei cada passo do algoritmo por meio de instantes, em que imprimi os valores de $L$ e $U$. No instante $0$ (não impresso), $L = I$ e $U = A$. Como $U[0]$ está acima da diagonal principal, ela se manterá como $U[0] = [4, -8, 1]$.

Para anular $U[1][0] = 6$, $m = 6/4 = 1.5$. Então, $L[1][0] = 1.5$ e $U[1] = [6, 5, 7] - 1.5 \times [4, -8, 1] = [6, 5, 7] - [6, -12, 1.5] = [6-6, 5-(-12), 7-1.5] = [0, 17, 5.5]$

Essas mudanças refletem o instante $1$ do algoritmo.

Como $U[2][0] = 0$, nenhuma mudança é necessária. $U[2] = U[2] - 0 \times U[0] = U[2]$. Ou seja, $L[2][0] = 0$. Tal iteração é expressa pelo instante $2$.

Para anular $U[2][1] = -10$, $m = \frac{-10}{17} \approx -0.58823529$. Então, $L[2][1] = -0.58823529$ e $U[2] = [0, -10, -3] - (-0.58823529) \times [0, 17, 5.5] \approx [0, -10, -3] + [0, 10, 3.23529412] = [0+0, -10+10, -
+3.23529412] = [0, 0, 0.23529412]$

Essas contas foram representadas pelo instante $3$, que mostra as mudanças finais do algoritmo.

Realizadas essas operações, determinei a decomposição LU a partir da matriz A. Então, obtive:

$$
L = \begin{bmatrix}
1  &        0 &          0  \\
1.5 &        1  &          0   \\
0   &      -0.58823529  &  1   \\
\end{bmatrix}
$$

e

$$
U = \begin{bmatrix}
4  &        -8 &          1  \\
0 &        17  &          5.5   \\
0   &      0  &  0.23529412   \\
\end{bmatrix}
$$